orttraining/orttraining/python/training/ortmodule.py

```diff
@@ -112,38 +115,44 @@ def _forward(self, *inputs, **kwargs):
             if not self._is_training():
                 return self._original_module(*inputs, **kwargs)
 
-            # Exporting module to ONNX for the first time
-            if not self._onnx_training:
+            # Exporting module to ONNX for the first time or when input changes
+            schema = _io._extract_schema({'args': copy.copy(inputs),
+                                          'kwargs': copy.copy(kwargs)})
+            build_gradient_graph = False
+            if not self._onnx_training or self._input_info.schema != schema:
+                build_gradient_graph = True
                 device_from_module = _utils.get_device_from_module(
                     self._original_module)
                 if not self._device or self._device != device_from_module:
                     self._device = device_from_module
                     if not self._device:
-                        raise RuntimeError(
-                            'A device must be specified in the model or data!')
-                self._get_inference_graph_and_init_gradient_graph_builder(
-                    *inputs, **kwargs)
-
-            # Flag to indicate whether the gradient_graph needs to be built
-            build_gradient_graph = self._current_input_shape is None
-            _, _, input_names_require_grad, new_input_shape = \
-                _ortmodule_io.parse_inputs_for_onnx_export(
-                    self._original_module_parameters, self._onnx_inference, *inputs, **kwargs)
-            initializer_names_to_train_set_user_model = {name for name, param in
-                self._flattened_output_module.named_parameters() if param.requires_grad}
+                        raise RuntimeError('A device must be specified by the model!')
+
+                self._onnx_inference = self._get_inference_graph(*inputs, **kwargs)
+                if self._save_onnx:
+                    onnx.save(self._onnx_inference, self._save_onnx_prefix + '_inference.onnx')
+                self._initialize_module_gradient_graph_builder()
+
+            input_info = _io.parse_inputs_for_onnx_export(self._original_module_parameters,
+                                                          self._onnx_inference,
+                                                          *inputs,
+                                                          **kwargs)
```

Comment:
Instead of passing `*inputs, **kwargs`, could we directly pass `inputs, kwargs` (and remove the corresponding `*` and `**` in the `parse_inputs_for_onnx_export()` signature)? My intuition here is that this will avoid a bunch of data structure shallow copies and dynamic allocations in the call path.